### PR TITLE
Improve tokenizer regex matching

### PR DESCRIPTION
**Pull request by mvorisek**
extracted from https://github.com/doctrine/sql-formatter/pull/119 as requested

Total runtime reduction is about -70%.

### CODE REVIEW

File: src/Tokenizer.php [hunk 1]
```diff
@@ -770,24 +770,24 @@ public function __construct()
         };
 
         // Set up regular expressions
-        $this->regexBoundaries       = '(' . implode(
+        $this->regexBoundaries       = '(?>' . implode(
```
Comment (greg0ire) [inline src/Tokenizer.php]: I don't know this notation. Where can I find docs on this?

File: src/Tokenizer.php [hunk 2]
```diff
@@ -897,7 +906,7 @@ private function createNextToken(string $string, string $upper, int $offset, Tok
         // Number (decimal, binary, or hex)
         if (
             preg_match(
-                '/\G(\d+(\.\d+)?|0x[\da-fA-F]+|0b[01]+)($|\s|"\'`|' . $this->regexBoundaries . ')/',
+                $this->nextTokenRegexNumber,
```
Comment (greg0ire) [inline src/Tokenizer.php:906]: So to be clear, the performance gain is about avoiding this kind of simple concatenation? Or is there more to it?

File: src/Tokenizer.php [hunk 3]
```diff
@@ -760,34 +761,86 @@ final class Tokenizer
...
+        self::$nextTokenRegexReservedNewline   = '/\G' . $regexReservedNewline . '(?=$|\s|' . $regexBoundaries . ')/';
+        self::$nextTokenRegexReserved          = '/\G' . $regexReserved . '(?=$|\s|' . $regexBoundaries . ')/';
+        self::$nextTokenRegexFunction          = '/\G' . $regexFunction . '(?=\s*\()/';
+        self::$nextTokenRegexNonReserved       = '/\G.*?(?=$|\s|["\'`]|' . $regexBoundaries . ')/';
+    }
+
+    /** @param list<string> $values */
+    private function makeRegexFromList(array $values, bool $sorted = false): string
+    {
+        // sort list by alphabet and from longest word to shortest
```
Comment (greg0ire) [inline src/Tokenizer.php]:
```suggestion
        // sort list alphabetically and from longest word to shortest
```

File: src/Tokenizer.php (hunk 1 above)
Comment (mvorisek) [reply src/Tokenizer.php]: We do not need/want to backtrack in the regular expressions, so we use "atomic group" - https://www.regular-expressions.info/atomic.html

File: src/Tokenizer.php (hunk 2 above)
Comment (mvorisek) [reply src/Tokenizer.php]: Let me know if you want to split this PR more. This is one main optimization, but see https://3v4l.org/9jitR/rfc#vgit.master_jit for possible maximal improvement of the improved regex generation from list. In the repro, I created/test worst case input, here in the commit messages, I tested one specific benchmark query only.

Also, the performance benefit of the last commit is huge when the Tokenizer is constructed more than once.

Comment (mvorisek) [reply src/Tokenizer.php:906]: I have split this PR now to focus only deduplicated regex generation once per Tokenizer lifetime.

**Comment by mvorisek** [comment]
I would be happy if this PR can be merged so I can work on another PR. Thank you.

File: src/Tokenizer.php (hunk 2 above)
Comment (mvorisek) [reply src/Tokenizer.php:906]: @greg0ire can I please have your review here?

**Comment by mvorisek** [comment]
@greg0ire can I please have your review here? Thank you.

**Comment by mvorisek** [comment]
@greg0ire can you please review this PR so I can move on with the followup PRs?

File: src/Tokenizer.php (hunk 2 above)
Comment (greg0ire) [reply src/Tokenizer.php:906]: … ok? There is still a diff here, so I don't understand why you're not answering my question.

Comment (mvorisek) [reply src/Tokenizer.php:906]: This PR focus on "Improve tokenizer regex matching" - which has 4 commits related to improving regex matching performance. If you want to furhter splut this PR, please let me know.

Comment (greg0ire) [reply src/Tokenizer.php:906]: I did not ask for a split though. I'm asking if the performance gain stems from avoiding that simple concatenation, or if there is more to it. Maybe I don't know, not concanenating helps PHP recognizing that this string never changes, and it compiles the regex only once or something?

Comment (mvorisek) [reply src/Tokenizer.php:906]: Each commit message summarizes the performance gain:

image

> avoiding that simple concatenation

The concatenation cannot be done on compile time, thus it is done on runtime and it needs allocation, per each "next token", which is of course very costly.

Comment (greg0ire) [reply src/Tokenizer.php:906]: > Each commit message summarizes the performance gain:

And that's great! I was just wondering how to explain the performance gain. If it's just avoiding the concatenation, then fine.

**Review by greg0ire — APPROVED** [review]

**Comment by mvorisek** [comment]
I would be happy if this PR can be merged so I can work on follow up PRs.

**Review by SenseException — APPROVED** [review]
Regex aren't easy to review. especially when they get concatinated with variables.

**Comment by greg0ire** [comment]
Thanks @mvorisek !

**Pull request merged**